main/provider/index.ts

```diff
@@ -448,33 +450,36 @@ export class Provider extends EventEmitter {
       const gas = gasFees(rawTx)
       const chainConfig = this.connection.connections['ethereum'][parseInt(rawTx.chainId)].chainConfig
 
-      const estimateGas = rawTx.gasLimit
-        ? Promise.resolve(rawTx)
-        : this.getGasEstimate(rawTx)
-            .then((gasLimit) => ({ ...rawTx, gasLimit }))
-            .catch((err) => {
-              approvals.push({
-                type: ApprovalType.GasLimitApproval,
-                data: {
-                  message: err.message,
-                  gasLimit: '0x00'
-                }
-              })
-
-              return { ...rawTx, gasLimit: '0x00' }
+      const estimateGasLimit = async () => {
+        if (rawTx.gasLimit) return rawTx.gasLimit
+        try {
+          return await this.getGasEstimate(rawTx)
+        } catch (error) {
+          if (error instanceof Error)
+            approvals.push({
+              type: ApprovalType.GasLimitApproval,
+              data: {
+                message: error.message,
+                gasLimit: '0x00'
+              }
             })
+          return '0x00'
+        }
+      }
 
-      estimateGas
-        .then((tx) => {
-          const populatedTransaction = populateTransaction(tx, chainConfig, gas)
+      const [gasLimit, recipientType] = await Promise.all([
+        estimateGasLimit(),
+        rawTx.to ? reveal.resolveEntityType(rawTx.to, parseInt(rawTx.chainId, 16)) : ''
+      ])
 
-          log.info({ populatedTransaction })
+      const tx = { ...rawTx, gasLimit, recipientType }
 
-          return populatedTransaction
-        })
-        .then((tx) => checkExistingNonceGas(tx))
-        .then((tx) => cb(null, { tx, approvals }))
-        .catch(cb)
+      try {
+        const populatedTransaction = populateTransaction(tx, chainConfig, gas)
+        cb(null, { tx: checkExistingNonceGas(populatedTransaction), approvals })
```

Comment:
do the nonce check on another line to prevent evaluating an operation on the return value line